### PR TITLE
Get panel overlap - gene annotation

### PR DESCRIPTION
**Pull request by keyvanelhami**
This pull request includes one snakemake file (Snakefile) and one python script (polishing_gene_list.py). The python script will polish and and cut the original reference file (refFlat), to a desired format that will be used by the Snakefile to merge and collapse the overlapping regions between the designed bed-file and the regions of interested from the refFlat gene list. For more information, see the comments in the beginning of each script.

To be able to test the scripts, edit the Snakefile by writing the  refFlat name in **genes** variable and the name if the bed-file without `.bed` extension in the **bedfile** variable

The output after testing the snakemake file should be like this ()

### CODE REVIEW

**Comment by hassanfa** [comment]
I'll review this tomorrow, but please move workflow to production/workflows/gene_annotation and either use all lower case for script name or camelcase:

Extract_regions_to_BedFile.py --> extract_regions_to_bedfile.py

I think the name doesn't reflect what it does, no? It is not extracting regions to bedfile, does it?

File: Extract_regions_to_BedFile.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    #If another variable name is used, a new column is created. If not using assign (), a series will be returned
+    #with only one column. The new df will be split() 'start' column values by ',' to a list.
+    #explode() will then paste transposed the 'start' column in "split_start" df and remaining columns will follow.
+    explode_end_df = input_df.assign(end=input_df['end'].str.split(',')).explode('end')
+    explode_end_df.reset_index(inplace=True)
+
+    return explode_end_df
+
+
+def grouping (input_df, strand):
```
Comment (hassanfa) [inline Extract_regions_to_BedFile.py]:
```suggestion
def group_transcript(input_df, strand):
```

Comment (hassanfa) [reply Extract_regions_to_BedFile.py]: `group_transcript` is more descriptive.

File: Extract_regions_to_BedFile.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    '''
+    input: df: DataFrame
+           path: string
+
+    output: csv-file
+
+    The output of this function will be tab-seperated csv file without any header.
+    '''
+
+    output_df = df.to_csv(outpath, sep='\t', header=False)
```
Comment (hassanfa) [inline Extract_regions_to_BedFile.py]: `to_csv` returns `None` no? If it is only this line, I don't think there is need for a function. Move to main.

File: Extract_regions_to_BedFile.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+@click.command()
+@click.option('--input', help = 'path to input file [required]')
+@click.option('--output', help = 'name and path for the output file [required]' )
+
+# The input and output file provided in the terminal will be passed into main() in read() and write_output() function.
+
+def main(input, output):
+
+    #Creating df from the input bed-file
+    read_df = read(str(input))
```
Comment (hassanfa) [inline Extract_regions_to_BedFile.py]:
```suggestion
    read_df = csv_to_df(str(input))
```

File: Extract_regions_to_BedFile.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+        #Adding new column where the gene name and transcript is merged together
+        file_df['gene_tx'] = file_df['gene'] + '_' + file_df['tx']
+
+    return file_df
+
+
+
+
+def strand_specific (input_df):
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply Extract_regions_to_BedFile.py]: `strand_specific` --> `extrac_strand_from_df` or something like that. It pays off in long run to be descriptive.

File: panel_overlap [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+rule all:
+    input:
+        'output/GMCKSolid4.1_hg19_mod_full.bed'
+
+# This rule will use python script (Extract_regions_to_BedFile.py) to reate a DataFrame from your
+# downloaded gene list from UCSC and then polish it to the desired csv-format: (chr, start, end, gene_tx_exon).
+# Open the python script (Extract_regions_to_BedFile.py) for more information. The cut command will cut the
+# col 2-4 and 7 and the perl command will remove the "chr" string from the file.
+
+rule create_gene_file:
```
Comment (hassanfa) [inline panel_overlap]: No version or log here?

**Comment by annaengstrom** [comment]
Would you like me to test this once more when the changes have been made? If yes, I'll promise I'll do it very quickly!

**Comment by hassanfa** [comment]
> Would you like me to test this once more when the changes have been made? I see there is also an added script. If yes, I'll promise I'll do it very quickly!

This is in a good state for testing, but lets wait until it is fully ready.

File: Extract_regions_to_BedFile.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
+#! /usr/bin/python
+
+"""
+Following script is taking UCSC gene list as input. See link below regarding how to extract the genes.
+Each gene has several transcripts where each row includes all the start and end pos of that transctipt.
+The below script will first read the file and then split it to two new df depending on strand (either + or -). Then
+split each start and end pos by ',' and adding the new values column-wise to a new dataframe. In the end, the
+two DataFrame with positive and negative strands will be merged. After fine tuning, the output will be a CSV-file.
+
+https://github.com/keyvanelhami/Extract_regions_from_panel-design
```
[the remark this answers is not in the document]
Comment (keyvanelhami) [reply Extract_regions_to_BedFile.py]: I have now: 
- Changed the name of the python and snakemake file.
- Moved the files to proper location.
- Modified the script to work with refFlat gene list

File: Extract_regions_to_BedFile.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+
+    return explode_end_df
+
+
+def grouping (input_df, strand):
+    '''
+    input: input_df: DataFrame
+           strand: string ("+", "-")
+
+    output: grouped_df: DataFrane
```
[the remark this answers is not in the document]
Comment (keyvanelhami) [reply Extract_regions_to_BedFile.py]: No it's a df, I will delete it from the description

File: Extract_regions_to_BedFile.py (hunk 2 above)
Comment (keyvanelhami) [reply Extract_regions_to_BedFile.py]: It dosen't remove `None`, or maybe I misunderstand you? I think it's better to have it as a function

File: panel_overlap (hunk 5 above)
Comment (keyvanelhami) [reply panel_overlap]: I don't really understand your comment here

File: Extract_regions_to_BedFile.py (hunk 3 above)
Comment (keyvanelhami) [reply Extract_regions_to_BedFile.py]: Done, please resolve when you have looked into it

File: Extract_regions_to_BedFile.py [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+    '''
+
+    col_header = ['tx', 'chr', 'strand', 'ExonCount', 'start', 'end', 'gene']
+
+    with open(path) as f:
+        #reading the csv file and creating a DataFrame
+        file_df = pd.read_csv(f, names=col_header, sep="\t")
+
+        #Adding new column where the gene name and transcript is merged together
+        file_df['gene_tx'] = file_df['gene'] + '_' + file_df['tx']
```
[the remark this answers is not in the document]
Comment (keyvanelhami) [reply Extract_regions_to_BedFile.py]: Done, please resolve when you have looked into it

File: Extract_regions_to_BedFile.py [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+
+    return file_df
```
[the remark this answers is not in the document]
Comment (keyvanelhami) [reply Extract_regions_to_BedFile.py]: Done, please resolve when you have looked into it

File: Extract_regions_to_BedFile.py [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+    The function will read an absolut path where the desired bed-file is located and will create
+    a new DataFrame which includes an extra column where the gene and transcript name are merged.
+
+    '''
+
+    col_header = ['tx', 'chr', 'strand', 'ExonCount', 'start', 'end', 'gene']
+
+    with open(path) as f:
+        #reading the csv file and creating a DataFrame
+        file_df = pd.read_csv(f, names=col_header, sep="\t")
```
[the remark this answers is not in the document]
Comment (keyvanelhami) [reply Extract_regions_to_BedFile.py]: Done, please resolve when you have looked into it

File: Extract_regions_to_BedFile.py [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+"""
+
+
+import pandas as pd
+import numpy as np
+import click
+
+
+
+def read(path):
```
[the remark this answers is not in the document]
Comment (keyvanelhami) [reply Extract_regions_to_BedFile.py]: Changed to `read_ucsc_refFlat`. Please resolve when you have looked into it

File: Extract_regions_to_BedFile.py [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+
+
+
+
+def strand_specific (input_df):
+
+    ''' Note: the header of the input file needs to be deleted before using this function
+    parameter (DataFrame)
+
+    output: DataFrame
```
[the remark this answers is not in the document]
Comment (keyvanelhami) [reply Extract_regions_to_BedFile.py]: Done, please resolve when you have looked into it

File: Extract_regions_to_BedFile.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (keyvanelhami) [reply Extract_regions_to_BedFile.py]: Done, please resolve when you have looked into it

File: Extract_regions_to_BedFile.py (hunk 1 above)
Comment (keyvanelhami) [reply Extract_regions_to_BedFile.py]: Done, please resolve when you have looked into it

**Comment by keyvanelhami** [comment]
@hassan please look into the changes I have made according to your comments and come back to me if there should be more changes.

File: utilities/gene_annotation/Snakefile [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+# This rule will first add a 4th column (which includes a ".") to the excisting bed-file. This is required
+# since the annotation with be on the 4th column on the final annotated bedfile. sed '1,3d' and the perl
+# command will remove the header and chr string from the chromosome column. E.g. "chr1" will be "1".
+
+rule modify_bed:
+    input:
+        '../../data_files/' + bedfile + '.bed'
+    output:
+        add_col = temp('output/' + bedfile + '.addcol.bed'),
+        rm_head_chr = ('output/' + bedfile + '.addcol.rmheader.rmchr.bed')
```
Comment (hassanfa) [inline utilities/gene_annotation/Snakefile]: Remove parentheses 
```suggestion
        rm_head_chr = 'output/' + bedfile + '.addcol.rmheader.rmchr.bed'
```

File: utilities/gene_annotation/Snakefile [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+    input:
+        all_genes=expand('{gene}', gene=target_bed),
+        panel_bed = 'output/' + bedfile + '.addcol.rmheader.rmchr.bed'
+    output:
+        overlap = temp('output/annotated.' + bedfile + '.overlap.bed'),
+        no_overlap = temp('output/annotated.' + bedfile + '.nooverlap.bed'),
+        sorted = temp('output/annotated.' + bedfile + '.sorted.overlap.bed'),
+        collapsed = temp('output/annotated.' + bedfile + '.collapsed.bed'),
+        annotated_bed = 'output/annotated.' + bedfile + '.bed'
+    log:
```
Comment (hassanfa) [inline utilities/gene_annotation/Snakefile]: This log is only capturing the last line. Either capture every single command's output or remove log.

File: utilities/gene_annotation/Snakefile [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+# string from the file.
+
+rule create_gene_file:
+    input:
+        python_script = "script/polishing_gene_list.py",
+        gene = expand('../reference/{genelist}', genelist=gene_list)
+    output:
+        python_output = temp(expand('output/{genelist}', genelist=gene_list)),
+        csv = ('output/'+ genes + '_final')
+    shell:
```
Comment (hassanfa) [inline utilities/gene_annotation/Snakefile]: Add version, just like there is one in `merge_all_gene_panels` or remove it.

i.e. all rules should have same behavior

File: utilities/gene_annotation/Snakefile [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+# since the annotation with be on the 4th column on the final annotated bedfile. sed '1,3d' and the perl
+# command will remove the header and chr string from the chromosome column. E.g. "chr1" will be "1".
+
+rule modify_bed:
+    input:
+        '../../data_files/' + bedfile + '.bed'
+    output:
+        add_col = temp('output/' + bedfile + '.addcol.bed'),
+        rm_head_chr = ('output/' + bedfile + '.addcol.rmheader.rmchr.bed')
+    shell:
```
Comment (hassanfa) [inline utilities/gene_annotation/Snakefile]: Add version, just like there is one in `merge_all_gene_panels` or remove it.

i.e. all rules should have same behavior

File: utilities/gene_annotation/script/polishing_gene_list.py [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+    The output of this function will be tab-seperated csv file without any header.
+    '''
+
+    output_df = df.to_csv(outpath, sep='\t', header=False)
+
+    return output_df
+# Using click package to provide the absolut path for the input and output file
+
+@click.command()
+@click.option('--input', help = 'path to input file [required]')
```
Comment (hassanfa) [inline utilities/gene_annotation/script/polishing_gene_list.py]: See here how to set an option required. And remove "[required]"

https://click.palletsprojects.com/en/7.x/options/#basic-value-options

File: utilities/gene_annotation/script/polishing_gene_list.py [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+@click.option('--output', help = 'name and path for the output file [required]' )
```
Comment (hassanfa) [inline utilities/gene_annotation/script/polishing_gene_list.py]: See here how to set an option required. And remove "[required]"

https://click.palletsprojects.com/en/7.x/options/#basic-value-options

File: utilities/gene_annotation/script/polishing_gene_list.py (hunk 17 above)
Comment (hassanfa) [reply utilities/gene_annotation/script/polishing_gene_list.py]: Input is a filename, so set it to path: https://click.palletsprojects.com/en/7.x/api/#click.Path

`type=click.Path()`

File: utilities/gene_annotation/script/polishing_gene_list.py (hunk 18 above)
Comment (hassanfa) [reply utilities/gene_annotation/script/polishing_gene_list.py]: Input is a filename, so set it to path: https://click.palletsprojects.com/en/7.x/api/#click.Path

`type=click.Path()`

File: Extract_regions_to_BedFile.py (hunk 2 above)
Comment (hassanfa) [reply Extract_regions_to_BedFile.py]: The function you have is just writing DF to a CSV. I'd prefer if it wasn't a function, but it's fine to have it this way.

And by `None` I meant this: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_csv.html Check the examples and return value's description.

Comment (hassanfa) [reply Extract_regions_to_BedFile.py]: You don't need to change anything here if you want to keep the function.

File: panel_overlap (hunk 5 above)
Comment (hassanfa) [reply panel_overlap]: The last rule has `version` and `log`. Either add version to all the rules or remove it from all.

About `log`: If you use the `log` then capture all the logs. In rule `merge_all_gene_panels` log is only capturing the last command's STDERR. Please read a bit on how you can capture multiple bash commands' STDERR and STDOUT.

Comment (keyvanelhami) [reply panel_overlap]: Removed log and version. I will read more about it before using them

File: utilities/gene_annotation/Snakefile (hunk 15 above)
Comment (keyvanelhami) [reply utilities/gene_annotation/Snakefile]: Removed versions and logs from all rules.

File: utilities/gene_annotation/Snakefile (hunk 16 above)
Comment (keyvanelhami) [reply utilities/gene_annotation/Snakefile]: Removed versions and logs from all rules.

File: utilities/gene_annotation/script/polishing_gene_list.py (hunk 17 above)
Comment (keyvanelhami) [reply utilities/gene_annotation/script/polishing_gene_list.py]: Added

File: utilities/gene_annotation/script/polishing_gene_list.py (hunk 18 above)
Comment (keyvanelhami) [reply utilities/gene_annotation/script/polishing_gene_list.py]: Added

**Comment by keyvanelhami** [comment]
@hassanfa all changed made, please look once more and see if it's ready to be merged to master

**Pull request merged**